htop: patch to (optionally) hide programme paths

nckx opened the pull request:
This makes using htop on NixOS a lot more pleasant (IMO).
Rebased patch has been merged upstream but not yet released.

peti commented:
Can we fetch that patch via HTTP from github instead of checking it into our repository?

nckx commented:
Not without me (re)creating the old branch, and even then: the [upstreamed patch](https://github.com/nckx/htop/commit/293eec4265e1a2555bffe351c5a85d6f55f1a52b) is functionally identical but heavily rebased.

jb55 commented:
:+1: I already have my local htop patched with this, would be great to see it in nixpkgs.

nckx commented:
So yes - in theory - but I'm not sure what this would save?

jb55 commented:
maybe just 

```
  patches = [fetchurl {
    url = https://github.com/nckx/htop/commit/293eec4265e1a2555bffe351c5a85d6f55f1a52b.diff
    sha256 = ...
  }]
```

saves us from having to save it in nixpkgs.

nckx commented:
@jb55: no, that's the rebased patch against `master` which is quite different from the `1.0.3` version currently in nixpkgs. The version against `1.0.3` is gone from my `htop` fork and only exists in this PR.

So yes, I could upload the patch to github (or anywhere), but then that applies to any patch file currently in nixpkgs.

peti commented:
If the patch cannot be applied to the current release version, how about switching our version of `htop` to the current HEAD of the upstream git repository? I'd really like to avoid creating a Nix-specific fork of this package by adding patches in a way upstream is unaware of.

nckx commented:
Oh for-- :-D The only reason I did it this way was to avoid exactly that. Git packages being frowned upon by some folks.

I love 'em. They have my patches.

So, all right (by you) if I just bump to HEAD and push?

jb55 commented:
One thing to keep in mind is the original author has refused to merge this patch, it would be great to have it as a configurable option at least.

nckx commented:
@jb55: since that would be me, I am highly sceptical of that claim.

peti commented:
@nckx, yes, bumping to HEAD is preferable to patching the 1.0.3 release, IMHO. Make sure to use the `fetchFromGithub` function so that we don't depend on `git` for downloading the source code, though.

jb55 commented:
@nckx wut? I'm referring to this issue https://github.com/hishamhm/htop/issues/1

nckx commented:
(@peti: wait, what? Just out of curiousity: are there *still* packages that do that?)

nckx commented:
@jb55: well, that's... interesting. I have no idea why upstream chose not to merge that.

nckx committed:
htop: 1.0.3 -> (git) 2015-08-12

Temporarily update to git HEAD for the "display program basename"
functionality it provides.

nbp commented:
And how would we switch the version to 1.1 when htop is upgraded and the patch is included / applied on top, if the version scheme is different?

Honestly, I would prefer us to default to the release version in Nixpkgs, and make a package option to switch to a different source.

vcunat commented:
Yeah, to achieve linearity, we would better name it `htop-1.1pre-2015-08-12`, for example (or whatever the next version might be called if not 1.1).

nckx commented:
Hence the "WIP" tag... @vcunat: no idea, and upstream is slow to respond (his good right). Does Nix have a scheme similar to `1.0.3-r2` in other distros?

peti commented:
I've changed the version string to `htop-1.0.3-186-gf2c053a` in 93b830f. That should remedy the `nix-env --update` issue.

vcunat commented:
@nckx: there's a detailed description of version ordering http://nixos.org/nix/manual/#ssec-version-comparisons. There's the special function of "pre" substring in there as the smallest element. But peti's solution should work the same in this case.

nckx commented:
The difference is that a `-pre$n` scheme requires one to ~~randomly guess~~ know the next version. `1.0.3b` (or `p` or whatever) would work but is ugly, so I too prefer peti's solution. Thanks!
